### PR TITLE
(Feat-CTA-17):Client_Creates_Feedback

### PR DESCRIPTION
**Pull request by Matthewtan9**
JIRA: link to jira ticket
[## Context:](https://champlainsaintlambert.atlassian.net/browse/CTA-17?atlOrigin=eyJpIjoiYzI3MDYzOGI3MDMzNDA5ZmE0MDc0NWJmOTQyNDBlNWMiLCJwIjoiaiJ9)
What is the ticket about and why are we doing this change?
- Changed some endpoints for Update QuoteRequest
- Finish adding the backend For add Feedback
- Finish testing backend
- Finish Frontend
- Finish Front end testing
- Change the placement for ratings for better visuals and added a star for more appeal.

## Before and After UI (Required for UI-impacting PRs)

Before 👍 
image

-Feedback form:
image
-Result after adding feedback :
image

-Admin/client Feedback bettered:
image

## Dev notes (Optional)
- Specific technical changes that should be noted
## Linked pull requests (Optional)
- pull request link

### CODE REVIEW

File: coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+  useEffect(() => {
+    // Fetching existing quote requests dates
+    axios.get("http://localhost:8080/feedbacks/dates", {
+      headers: {
+        // @ts-ignore
+        'X-XSRF-TOKEN': auth.getXsrfToken()
+      }
+    })
+      .then(response => {
+        setExistingFeedBacks(response.data);
```
Comment (drallo22) [inline coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: What is the reason for the functionality of getting all the feedback dates? I see you call the backend method and then set the existing feedbacks states but then it is never used?

File: coupetonarbrebackend/src/main/java/com/example/coupetonarbrebackend/Feedback/PresentationLayer/FeedbackController.java [hunk 2]
```diff
@@ -42,11 +43,29 @@ public ResponseEntity<List<FeedbackResponseDTO>> getAllFeedbacksClient(@Authenti
...
     @PreAuthorize("hasAuthority('Admin') || hasAuthority('Client')")
     @GetMapping("/{feedbackId}")
-    public ResponseEntity<FeedbackResponseDTO> getFeedbackById(@PathVariable String feedbackId){
+    public ResponseEntity<FeedbackResponseDTO> getFeedbackById(@PathVariable String feedbackId) {
         return ResponseEntity.ok().body(feedbackService.getFeedbackById(feedbackId));
     }
 
+    @PreAuthorize("hasAuthority('Client')")
+    @PostMapping("/client")
+    public  ResponseEntity<FeedbackResponseDTO> clientAddFeedbacks(@AuthenticationPrincipal OidcUser principal,
```
Comment (drallo22) [inline coupetonarbrebackend/src/main/java/com/example/coupetonarbrebackend/Feedback/PresentationLayer/FeedbackController.java:55]: Just a small thing for consistency, feedback should not have an s

File: coupetonarbrebackend/src/main/java/com/example/coupetonarbrebackend/Feedback/PresentationLayer/FeedbackController.java [hunk 3]
```diff
@@ -42,11 +43,29 @@ public ResponseEntity<List<FeedbackResponseDTO>> getAllFeedbacksClient(@Authenti
...
+        String clientId = principal.getSubject();
+        log.info("Client with clientId: {}", clientId);
+
+        FeedbackResponseDTO feedbackResponseDTO = feedbackService.clientAddFeedbacks(feedbackRequestDTO, clientId);
+        return ResponseEntity.status(HttpStatus.CREATED).body(feedbackResponseDTO);
+    }
+
+    @PreAuthorize("hasAuthority('Client')")
+    @GetMapping("/dates")
+    public ResponseEntity<List<FeedbackDatesDTO>> getAllFeedbackDates(){
```
Comment (drallo22) [inline coupetonarbrebackend/src/main/java/com/example/coupetonarbrebackend/Feedback/PresentationLayer/FeedbackController.java:66]: Why is this necessary?

**Review by drallo22 — COMMENTED** [review]
I just had some questions regarding the feedback dates methods

File: coupetonarbrebackend/src/main/java/com/example/coupetonarbrebackend/Feedback/PresentationLayer/FeedbackController.java (hunk 2 above)
Comment (Matthewtan9) [reply coupetonarbrebackend/src/main/java/com/example/coupetonarbrebackend/Feedback/PresentationLayer/FeedbackController.java:55]: Yes will remove it for consistency.

File: coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx (hunk 1 above)
Comment (Matthewtan9) [reply coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: While the direct use of the fetched data may not be visible in the provided code snippet. The existing feedback dates are used for date validation in the feedback creation form, ensuring users can only select valid dates and excluding certain dates (e.g., holidays).

File: coupetonarbrebackend/src/main/java/com/example/coupetonarbrebackend/Feedback/PresentationLayer/FeedbackController.java (hunk 3 above)
Comment (Matthewtan9) [reply coupetonarbrebackend/src/main/java/com/example/coupetonarbrebackend/Feedback/PresentationLayer/FeedbackController.java:66]: As seen how the other services uses, Im guessing its used for the date validation in the feedback. Since feedback when creating one, uses date as one of its field.

**Review by ebond07 — COMMENTED** [review]
I'll approve this as soon as the comments are resolved from David.

File: coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx (hunk 1 above)
Comment (drallo22) [reply coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: I think ChatGPT might be wrong on this one because the fetched data is not being used anywhere not even for the validation. Also, there shouldn't be any date validation for feedback and the dates being retrieved has nothing to do with blocking out holidays. In the other services I get the dates and use those dates and times to ensure that there cannot be two quotes or appointments at the same time, however, that does not apply to feedback since it does not matter when feedback is left. This goes with my other question which is what is the date  that the user is inputting? the date where their appointment was done? Either way, the only validation there would be is that they can only select the date where their appointment took place. For that to happen, you can do that the user must select one of their already completed appointments which would be linked to the feedback. Either way if you decide not to go that route, the retrieval of the dates is not necessary because it is not being used, and even if it was it shouldn't be done.

Comment (Matthewtan9) [reply coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: Should I simply delete it and set the date to the pre-filled current date when the user submits the feedback?

Comment (drallo22) [reply coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: Yeah that makes sense but that was one of my questions, what date is supposed to be inputted? Because yes it makes sense for the admin to see when the feedback was sent but he also needs to see what appointment or date it was for

Comment (dazhuox) [reply coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: Wait holdup, why does the user need to input a date? Cant it just be the current datetime of the feedback creation? And David, since a Feedback entity contains an appointmentId, I think that's already covered, there's just maybe a button that redirects the user from this page to the appointments page missing?

Comment (Matthewtan9) [reply coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: Honestly there was not really any thougth process over the date input. I can just remove and make it do the date is just prefilled with the date of the creation.

Comment (drallo22) [reply coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: There doesn't need to be a date inputted if there is an appointment linked to it. In the screenshots and code I can't find where the appointmentId is linked to the feedback which is why I'm asking. Plus in the screenshots it says appointmentId but it is empty so I can't tell if it is implemented/working

Comment (Matthewtan9) [reply coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: The appointment Id is currently null right now because we discussed in the call how to implement the validation first. I set it to null since I'm not sure I'm supposed to use it for now. That is why I have only the basic add function currently without using the appointment Id.

Comment (dazhuox) [reply coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: Oh you're right, I didn't spot that appointmentId wasn't showing in the getById.

So Matthew I think the stuff to change would be the date and making the appointmentId appear in the getById

Comment (Matthewtan9) [reply coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: I know but every time I remove the null and set it back to not null, the create does not work anymore.

Comment (drallo22) [reply coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: You would need to create a method that gets the clients approved appointments, you would then display them in a dropdown so that the user can select which one they wanna base their feedback on (in case they have multiple), then when a user selects one it gets it byId and then you set the appointmentId in the request to that ID. It is not hard or long, but if you wanna implement that in your other PR that's fine

Comment (Matthewtan9) [reply coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: Is that both backend and front end ?

Comment (drallo22) [reply coupetonarbrefrontend/src/pages/CreateFeedbackClient.jsx:124]: Yeah
